driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -181,23 +181,26 @@ public void onResult(final List<Document> result, final Throwable t) {
     private void processResults(final int previousAmountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
         try {
             int amountRead = previousAmountRead;
-            while (currentPosition < fileInfo.getLength() && dst.remaining() > 0 && !resultsQueue.isEmpty()) {
-                if (buffer == null || bufferOffset == buffer.length) {
+            int amountToCopy = dst.remaining();
+            while (currentPosition < fileInfo.getLength() && amountToCopy > 0) {
```

Comment:
This loop condition is confusing.  It checks that amountToCopy > 0, but amountToCopy is not updated after that amount is copied into the destination buffer on line 198.  Just to check, I added 

     amountToCopy = dst.remaining();

below line 201 and the tests still pass.  So maybe entering the loop one extra time is harmless, but it looks wrong.